rmw_connext_shared_cpp/src/node_names.cpp

```diff
@@ -90,45 +108,93 @@ get_node_names(
       std::vector<uint8_t> kv(data, data + pbtd.user_data.value.length());
+
       auto map = rmw::impl::cpp::parse_key_value(kv);
+
       auto name_found = map.find("name");
       auto ns_found = map.find("namespace");
 
       if (name_found != map.end()) {
         name = std::string(name_found->second.begin(), name_found->second.end());
       }
-      if (name_found != map.end()) {
+
+      if (ns_found != map.end()) {
         namespace_ = std::string(ns_found->second.begin(), ns_found->second.end());
       }
+
       if (name.empty()) {
         // use participant name if no name was found in the user data
         if (pbtd.participant_name.name) {
           name = pbtd.participant_name.name;
         }
       }
     }
+
+    // ignore discovered participants without a name
     if (name.empty()) {
-      // ignore discovered participants without a name
-      node_names->data[i] = nullptr;
-      node_namespaces->data[i] = nullptr;
+      tmp_names_list.data[i] = nullptr;
+      tmp_namespaces_list.data[i] = nullptr;
       continue;
     }
 
-    node_names->data[i] = rcutils_strdup(name.c_str(), allocator);
-    if (!node_names->data[i]) {
-      RMW_SET_ERROR_MSG("could not allocate memory for node name");
-      goto fail;
+    tmp_names_list.data[named_nodes_num] = rcutils_strdup(name.c_str(), allocator);
+    if (!tmp_names_list.data[named_nodes_num]) {
+      RMW_SET_ERROR_MSG("could not allocate memory for a node's name");
+      final_ret = rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
+      goto cleanup;
     }
 
-    node_namespaces->data[i] = rcutils_strdup(namespace_.c_str(), allocator);
-    if (!node_namespaces->data[i]) {
-      RMW_SET_ERROR_MSG("could not allocate memory for node namespace");
-      goto fail;
+    tmp_namespaces_list.data[named_nodes_num] = rcutils_strdup(namespace_.c_str(), allocator);
+    if (!tmp_namespaces_list.data[named_nodes_num]) {
+      RMW_SET_ERROR_MSG("could not allocate memory for a node's namespace");
+      final_ret = rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
+      goto cleanup;
     }
+
+    ++named_nodes_num;
+  }
+
+  // prepare actual output without empty names
+  rcutils_ret = rcutils_string_array_init(node_names, named_nodes_num, &allocator);
+  if (rcutils_ret != RCUTILS_RET_OK) {
+    RMW_SET_ERROR_MSG("could not allocate memory for node_names output");
+    final_ret = rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
+    goto cleanup;
+  }
+
+  rcutils_ret = rcutils_string_array_init(node_namespaces, named_nodes_num, &allocator);
+
+  if (rcutils_ret != RCUTILS_RET_OK) {
+    RMW_SET_ERROR_MSG("could not allocate memory for node_namespaces output");
+    final_ret = rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
+    goto cleanup;
+  }
+
+  for (auto i = 0; i < named_nodes_num; ++i) {
+    node_names->data[i] = rcutils_strdup(tmp_names_list.data[i], allocator);
+    tmp_names_list.data[i] = nullptr;
```

Comment:
isn't this leaking memory?

Reply:
Yes, I suppose it does. Good catch! I don't think this line or the other at L178 are necessary.

See https://github.com/ros2/rmw_connext/pull/395 for a fix.